### PR TITLE
Implement generational storage with `SELFDESTRUCT` tests

### PR DESCRIPTION
**Pull request by mfornet**
Check state is removed after selfdestruct is called following this description: 

`https://gist.github.com/mfornet/a8ffcedfffd22612e34cf0b3c3f5c9e3`

@joshuajbouw use this test, to check that reset state is working.

----

Also add a test (`test_self_destruct_with_submit`) where function that call `selfdestruct` is called using `engine.submit` rather than `engine.call` and it fails with:
```
`Result::unwrap()` on an `Err` value: InconsistentStateError(IntegerOverflow)'
```
This was found by chance. @birchmd @artob can you take a look at this.

### CODE REVIEW

**Comment by joshuajbouw** [comment]
Hm, getting an error that I'm missing a file? Can you double-check that it's pushed?

```rust
    pub fn load() -> Self {
        Self(solidity::ContractConstructor::compile_from_extended_json(
            "etc/eth-contracts/artifacts/contracts/StateTest.sol/SelfDestructFactory.json",
        ))
    }
```

**Comment by birchmd** [comment]
> Hm, getting an error that I'm missing a file?

Looks like it's supposed to be generated by `yarn` (see changes to `Makefile`). We'll just need to update the GitHub action to run that first.

**Comment by birchmd** [comment]
By the way @mfornet is `yarn` (or hardhat or whatever) needed to compile this contract for some reason? If this existing `solidity::compile` function would work (as is used for the standard precompiles test) then I would prefer that to adding an additional tooling dependency.

Even if `yarn` is needed we could probably use a similar tactic to what I used with the standard precompiles contract to get rust to call `yarn` itself rather than have it go through the `Makefile`. This would be a better way than what I suggested about about changing the GitHub action.

File: src/test_utils/mod.rs
Comment (birchmd) [inline src/test_utils/mod.rs:33]: Nice! I like this a lot. Might be worth refactoring other tests to use this in the future.

**Comment by birchmd** [comment]
> called using `engine.submit` rather than `engine.call` and it fails

Noted. I'll look into it.

**Comment by joshuajbouw** [comment]
Got it now @birchmd

**Comment by birchmd** [comment]
@mfornet the error you encountered with `submit` is fixed in #85

**Comment by artob** [comment]
@joshuajbouw @mfornet Please resolve the conflicts now after the merge of #87.

File: src/precompiles/mod.rs
Comment (joshuajbouw) [inline src/precompiles/mod.rs:213]: Better version of this would be to use `copy_from_slice`

**Comment by joshuajbouw** [comment]
Just merged `develop` into this branch carefully as there were at least 20 file conflicts.

**Review by joshuajbouw — APPROVED** [review]
Generally fine, I still have to wait for the CI to go through its run, but my only complaint is that there is obviously quite a bit of feature creep going on here. Do be very careful with that going forward. Would be best to have each of those as separate PRs. Also, do thoroughly document all changes that had been made in the description.

File: src/precompiles/mod.rs
Comment (mfornet) [reply src/precompiles/mod.rs]: Is it possible to use `copy_from_slice` in `const fn`? I think that's why it was done this way.

Comment (birchmd) [reply src/precompiles/mod.rs]: Indeed `copy_from_slice` is not allowed (because `&mut` pointers are not allowed). See [here](https://github.com/aurora-is-near/aurora-engine/pull/107#discussion_r640885291) for further discussion on an implementation which has less boilerplate, but it worse in my opinion.

**Comment by mfornet** [comment]
Found a subtle bug in the [way keys were created](https://github.com/aurora-is-near/aurora-engine/pull/84/commits/27314b19303d10934a1cb00b8d9812ea6344ceec#diff-18c17f16d6477d10ed90360fe38e0e13fc117ecd9548f8f9374236c3ec7f8750L106-R107). Some tests are still failing (see CI), and is a regression introduced in PR #87, already checked on previous commit that it was working fine. 

I was looking into it, but still haven't figured out the issue.

[Simple test where it fails](https://github.com/aurora-is-near/aurora-engine/blob/27314b19303d10934a1cb00b8d9812ea6344ceec/src/tests/erc20.rs#L15)

1. A standard ERC20 implementation (form OpenZeppelin) is deployed.
2. Check balance from a user (it is 0)
3. Mint 10 tokens for the user
4. Check balance from a user (it is 10)

At step 3 the transaction fails.

File: src/engine.rs
Comment (mfornet) [inline src/engine.rs:585]: We need to understand what is the meaning of the flag `reset_storage`. From [this comment](https://github.com/aurora-is-near/sputnikvm/blob/master/src/backend/mod.rs#L38-L40) it seems that the storage needs to be wiped, however, as seen in some tests, this is not working as expected.

Maybe the comment in sputnikvm refers to a different storage 🤔

Comment (birchmd) [reply src/engine.rs:585]: I see the error. `remove_all_storage` increments the generation by 1, but then in the storage iterator we are using the previous generation. We should do the following:

```
if reset_storage {
    Engine::remove_all_storage(&address, generation);
    generation += 1;
}
```

Comment (birchmd) [reply src/engine.rs:585]: I confirmed all tests pass with that change, but actually, that solution is not so good because it interacts poorly with the `if delete_empty` block later. Need to think about it more carefully.

Comment (birchmd) [reply src/engine.rs]: Fixed in https://github.com/aurora-is-near/aurora-engine/pull/84/commits/dc49d44e224e5038bcf1a44bbd00c11fdbf80095

**Review by birchmd — APPROVED** [review]
Great work team!

**Pull request merged**
